dag_in_context/src/greedy_dag_extractor.rs

```diff
@@ -329,14 +344,72 @@ impl<'a> Extractor<'a> {
             self.termdag.app(op.into(), children)
         };
 
-        self.correspondence
-            .entry(term.clone())
-            .or_default()
-            .insert(node_id);
+        Self::add_correspondence(&mut self.correspondence, term.clone(), node_id.clone());
 
         term
     }
 
+    fn add_correspondence(
+        correspondence: &mut IndexMap<Term, NodeId>,
+        term: Term,
+        node_id: NodeId,
+    ) {
+        if let Some(existing) = correspondence.insert(term.clone(), node_id.clone()) {
+            assert_eq!(existing, node_id);
+        }
+    }
+
+    /// Add `term` to `current_costs`, returning
+    /// 1) a new term that takes advantage of sharing
```

Comment:
An example in the comment for the first bullet point can be helpful for understanding, e.g., for E-node `Add(c1, Op(c1))`, The two child cost sets of this E-node can have different witnesses for `c1` and this tries to fix it up by having them agreeing on witnesses.

Reply:
added example